<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="qtrobot">

	<xacro:property name="radius_two" value="0.02" />
	<xacro:property name="radius_three" value="0.03" />
	<xacro:property name="elbow_roll_length" value="0.04" />
	<xacro:property name="shoulder_pitch_length" value="0.05" />
	<xacro:property name="shoulder_roll_length" value="0.1" />
	<xacro:property name="hand_length" value="0.2" />

	<xacro:macro name="elbow_origin">
	    <origin rpy="0 1.5708 0" xyz="0 0 0"/>
	</xacro:macro>

	<xacro:macro name="shoulder_origin">
		<origin rpy="1.5708 0 0" xyz="0 0.03 0"/>
	</xacro:macro>

	<xacro:macro name="hand_origin">
	    <origin rpy="0 1.5708 0" xyz="-0.1 0 0"/>
	</xacro:macro>

	<xacro:macro name="shoulderPitch_link" params="position">
	  <link name="${position}ShoulderPitch">
	    <visual>
	      <geometry>
	        <cylinder length="${shoulder_pitch_length}" radius="${radius_three}"/>
	      </geometry>
	      <xacro:shoulder_origin />
	      <material name="white"/>
	    </visual>
	  </link>
	</xacro:macro>

	<xacro:macro name="shoulderRoll_link" params="position">
		<link name="${position}ShoulderRoll">
			<visual>
			  	<geometry>
			    	<cylinder length="${shoulder_roll_length}" radius="${radius_three}"/>
			  	</geometry>
			    <xacro:shoulder_origin />
			  	<material name="white"/>
			</visual>
		</link>
	</xacro:macro>

	<xacro:macro name="elbowRoll_link" params="position">
		<link name="${position}ElbowRoll">
			<visual>
			  	<geometry>
			    	<cylinder length="${elbow_roll_length}" radius="${radius_two}"/>
			  	</geometry>
			  	<xacro:elbow_origin />
			  	<material name="white"/>
			</visual>
		</link>
	</xacro:macro>

	<xacro:macro name="hand_link" params="position">
		<link name="${position}Hand">
		    <visual>
		      	<geometry>
		       		<cylinder length="${hand_length}" radius="${radius_three}"/>
		      	</geometry>
		      	<xacro:hand_origin />
		    </visual>
		</link>
	</xacro:macro>

	<material name="white">
		<color rgba="1 1 1 1"/>
	</material>

	<link name="base_link">
		<visual>
		  	<geometry>
		    <!-- <box size="0.10 0.15 0.43"/> -->
		    	<cylinder length="0.43" radius="0.10"/>
		 	</geometry>
		  	<origin rpy="0 0 0" xyz="0 0 0.1685"/>
		  	<material name="white"/>
		</visual>
	</link>

	<!-- Head -->
	<link name="HeadYaw">
		<visual>
		  	<geometry>
		    	<cylinder length="0.04" radius="${radius_two}"/>
		  	</geometry>
		  	<origin rpy="0 0 0" xyz="0 0 0"/>
		  	<material name="white"/>
		</visual>
	</link>

	<link name="HeadPitch">
		<visual>
			<geometry>
		    	<sphere radius="0.15"/>
		  	</geometry>
		  	<origin rpy="1.5708 0 0" xyz="0 0 0.06"/>
		  	<material name="white"/>
		</visual>
	</link>

	<joint name="HeadYaw" type="revolute">
		<parent link="base_link"/>
		<child link="HeadYaw"/>
		<origin rpy="0 0 0" xyz="0 0 0.338"/>
		<axis xyz="0 0 1"/>
		<limit effort="1000" lower="-1.0472" upper="1.0472" velocity="100"/>
	</joint>

	<joint name="HeadPitch" type="revolute">
		<parent link="HeadYaw"/>
		<child link="HeadPitch"/>
		<origin rpy="0 0 0" xyz="0 0 0.0962"/>
		<axis xyz="0 1 0"/>
		<limit effort="1000" lower="-0.261799" upper="0.436332" velocity="30"/>
	</joint>

	<!-- Right Arm -->
	<xacro:shoulderPitch_link position="Right"/>
	<xacro:shoulderRoll_link position="Right"/>
	<xacro:elbowRoll_link position="Right"/>
	<xacro:hand_link position="Right"/>

	  <joint name="RightShoulderPitch" type="revolute">
	    <parent link="base_link"/>
	    <child link="RightShoulderPitch"/>
	    <origin rpy="3.14159 0 0" xyz="0 -0.08 0.396"/>
	    <axis xyz="0 1 0"/>
	    <limit effort="1000" lower="-2.44346" upper="2.44346" velocity="100"/>
	  </joint>

	  <joint name="RightShoulderRoll" type="revolute">
	    <parent link="RightShoulderPitch"/>
	    <child link="RightShoulderRoll"/>
	    <origin rpy="0 -1.5708 0" xyz="0 0.0445 0"/>
	    <axis xyz="1 0 0"/>
	    <limit effort="1000" lower="-1.309" upper="0.122173" velocity="100"/>
	  </joint>

	  <joint name="RightElbowRoll" type="revolute">
	    <parent link="RightShoulderRoll"/>
	    <child link="RightElbowRoll"/>
	    <origin rpy="0 0 0" xyz="0 0.07708 0"/> -->
	    <axis xyz="1 0 0"/>
	    <limit effort="1000" lower="-1.5708" upper="-0.122173" velocity="100"/>
	  </joint>

	  <joint name="RightHand" type="fixed">
	    <parent link="RightElbowRoll"/>
	    <child link="RightHand"/>
	    <origin rpy="-1.5708 0 1.5708" xyz="0 0.184 0"/>
	  </joint>

	<!-- Left Arm -->
	<xacro:shoulderPitch_link position="Left"/>
	<xacro:shoulderRoll_link position="Left"/>
	<xacro:elbowRoll_link position="Left"/>
	<xacro:hand_link position="Left"/>

	<joint name="LeftShoulderPitch" type="revolute">
	    <parent link="base_link"/>
	    <child link="LeftShoulderPitch"/>
	    <origin rpy="0 0 0" xyz="0 0.08 0.396"/> -->
	    <axis xyz="0 1 0"/>
	    <limit effort="1000" lower="-2.44346" upper="2.44346" velocity="100"/>
	</joint>

	<joint name="LeftShoulderRoll" type="revolute">
	    <parent link="LeftShoulderPitch"/>
	    <child link="LeftShoulderRoll"/>
	    <origin rpy="0 -1.5708 0" xyz="0 0.0445 0"/>
	    <axis xyz="1 0 0"/>
	    <limit effort="1000" lower="-1.309" upper="0.122173" velocity="100"/>
	</joint>

	<joint name="LeftElbowRoll" type="revolute">
	    <parent link="LeftShoulderRoll"/>
	    <child link="LeftElbowRoll"/>
	    <origin rpy="0 0 0" xyz="0 0.07708 0"/>
	    <axis xyz="1 0 0"/>
	    <limit effort="1000" lower="-1.5708" upper="-0.122173" velocity="100"/>
	</joint>

	<joint name="LeftHand" type="fixed">
	    <parent link="LeftElbowRoll"/>
	    <child link="LeftHand"/>
	    <origin rpy="1.5708 0 1.5708" xyz="0 0.184 0"/>
	</joint>

</robot>